### PR TITLE
i#1312 AVX-512 support: Extend priv_mcontext_t.simd to AVX-512 number of registers.

### PR DESCRIPTION
**Pull request by hgreving2304**
Adds to DynamoRIO's fixed number of SIMD slots to include AVX-512 registers on 64-bit
UNIX. If processor and OS support AVX-512, proc_num_simd_registers() will now return 32.
The number of registers has not been updated in the Windows build yet.

Please note that proc_num_simd_saved() is not yet switched to reflect AVX-512, even if
AVX-512 instructions are present in the application. DynamoRIO does not yet maintain,
preserve, or restore AVX-512 context yet.

Fixes corresponding tests and adds comments.

Further adds to backwards compatibility handling in dr_mcontext_to_priv_mcontext() and
priv_mcontext_to_dr_mcontext(). It is now accepting the simd structure's layout before
this commit. This patch adds more error handling if the wrong size is passed by the
client.

Issue: #1312

### CODE REVIEW

File: core/arch/arch.c [hunk 1]
```diff
@@ -3383,17 +3383,42 @@ dr_mcontext_to_priv_mcontext(priv_mcontext_t *dst, dr_mcontext_t *src)
...
 #ifdef X86
             if (src->size > offsetof(dr_mcontext_t, simd)) {
-                if (src->size > offsetof(dr_mcontext_t, simd) +
-                        MCXT_NUM_SIMD_SSE_AVX_SLOTS * YMM_REG_SIZE) {
+                if (MCXT_NUM_SIMD_SLOTS > MCXT_NUM_SIMD_SSE_AVX_SLOTS &&
+                    src->size > offsetof(dr_mcontext_t, simd) +
+                            MCXT_NUM_SIMD_SSE_AVX_SLOTS * ZMM_REG_SIZE) {
+                    if (src->size < offsetof(dr_mcontext_t, simd) + sizeof(dst->simd))
+                        return false;
+                    /* Full copy, UNIX 64-bit. XXX i#1312: We don't support AVX-512
```
Comment (derekbruening) [inline core/arch/arch.c]: Confused: isn't the full copy covered by the == case up front?  So why would this be "full copy"?

File: core/arch/arch.c [hunk 2]
```diff
@@ -3383,17 +3383,42 @@ dr_mcontext_to_priv_mcontext(priv_mcontext_t *dst, dr_mcontext_t *src)
...
+                     * extended number of registers in 64-bit Windows yet.
+                     */
+                    memcpy(&dst->simd, &src->simd,
+                           MCXT_NUM_SIMD_SSE_AVX_SLOTS * ZMM_REG_SIZE);
+                } else if (MCXT_NUM_SIMD_SLOTS == MCXT_NUM_SIMD_SSE_AVX_SLOTS &&
+                           src->size > offsetof(dr_mcontext_t, simd) +
+                                   MCXT_NUM_SIMD_SSE_AVX_SLOTS * YMM_REG_SIZE) {
+                    if (src->size < offsetof(dr_mcontext_t, simd) + sizeof(dst->simd))
+                        return false;
+                    /* Full copy if build is not UNIX 64-bit. */
```
Comment (derekbruening) [inline core/arch/arch.c]: Again, wouldn't a full copy be the up-front == case?

File: core/arch/x86/x86_asm_defines.asm [hunk 3]
```diff
@@ -49,7 +49,7 @@
 # ifdef WINDOWS
 #  define MCXT_NUM_SIMD_SLOTS 6 /* xmm0-5 */
 # else
-#  define MCXT_NUM_SIMD_SLOTS 16 /* xmm0-15 */
+#  define MCXT_NUM_SIMD_SLOTS 32 /* xmm0-15 */
```
Comment (derekbruening) [inline core/arch/x86/x86_asm_defines.asm]: Update the comment

File: core/arch/x86_code_test.c [hunk 4]
```diff
@@ -134,7 +134,10 @@ unit_test_get_ymm_caller_saved()
     register __m256 ymm15 asm("ymm15");
 #        endif
 
-    for (int regno = 0; regno < proc_num_simd_registers(); ++regno) {
+    /* The function get_ymm_caller_saved is intended to be used for AVX (no AVX-512). It
+     * currently doesn't cover extended AVX-512 registers.
```
Comment (derekbruening) [inline core/arch/x86_code_test.c]: "currently" implies it should and will be changed: so TODO i#?

File: core/arch/arch.c (hunk 1 above)
Comment (hgreving2304) [reply core/arch/arch.c]: The language is misleading. I meant to say that we're copying the entire sizeof(simd) in these cases. I am re-wording the comment.

File: core/arch/arch.c (hunk 2 above)
Comment (hgreving2304) [reply core/arch/arch.c]: See above.

File: core/arch/x86_code_test.c (hunk 4 above)
Comment (hgreving2304) [reply core/arch/x86_code_test.c]: No, it should stay. I don't see a use case for ymm16-ymm31 right now? That's why I added the "currently". I'll remove the adverb.

File: core/arch/x86/x86_asm_defines.asm (hunk 3 above)
Comment (hgreving2304) [reply core/arch/x86/x86_asm_defines.asm]: Done.

File: core/arch/x86_code_test.c (hunk 4 above)
Comment (hgreving2304) [reply core/arch/x86_code_test.c]: Done.

File: core/arch/arch.c (hunk 1 above)
Comment (hgreving2304) [reply core/arch/arch.c]: Re-phrased/clarified.

File: core/arch/arch.c (hunk 2 above)
Comment (hgreving2304) [reply core/arch/arch.c]: Re-phrased/clarified.

**Pull request merged**